Add release action

palewire opened the pull request:
Here you find a basic implementation of a release action. When a Github Release is triggered, it will run after the successful completion of the of the lint and test steps. [In my testing](https://github.com/palewire/npm-ci-test), it is successfully uploading an empty package to my personal npmjs.com account.

However, it may need to be customized to support the `yarn build` rollup step in this repository. Rather than try to guess how you'd like that part to work, I thought I'd put this up here as a starting point for discussion. Let me know what more I can do to improve the proposal.

hobbes7878 commented:
Cool. Just outlining the release job to make sure I get it: We add a git tag for new patch/minor/major on master. That then bootstraps and builds (missing rn, as you said) the library. Finally, npm versions and publishes to the registry with the tag we pushed (so need to consistently tag as `1.0.1` not `v1.0.0`, I think??).

Other Qs: Do we need to restrict this job to just master branch and just our repo ([e.g.](https://github.com/sveltejs/kit/blob/master/.github/workflows/release.yml#L12))?
Do we need to commit back to the repo, since npm-version is updating package.json with the new version number? Would it make sense to throw release in its own yaml restricted to push on master?

palewire commented:
> Cool. Just outlining the release job to make sure I get it: We add a git tag for new patch/minor/major on master. That then bootstraps and builds (missing rn, as you said) the library. Finally, npm versions and publishes to the registry with the tag we pushed (so need to consistently tag as `1.0.1` not `v1.0.0`, I think??).

Yes, that's basically it. Though it should stamp the version with whatever you pass into the release name in GitHub's web interface.

> Other Qs: Do we need to restrict this job to just master branch and just our repo ([e.g.](https://github.com/sveltejs/kit/blob/master/.github/workflows/release.yml#L12))?

The release event that qualifies to run new final step of the action should only get triggered for the default branch, I think. We want the tests that precede it to run for every push to every branch, so I wouldn't want to add a workflow-level branch restriction. I could imagine adding another statement to the `if` clause if you can imagine a circumstance where a tagged release would run for another branch.

The alternative method for handling the event, which is [recommended by the GitHub docs](https://docs.github.com/en/actions/publishing-packages/publishing-nodejs-packages), is to craft a separate release action that runs on the `published` hook that fires after a GitHub Release is created. I tend not to use that approach because I like having the linter and test jobs act as hurdles to prevent broken code from shipping, and I prefer not to replicate those jobs in two different actions. That said, this alternative structure totally works and I'd be happy to switch to using it here.

> Do we need to commit back to the repo, since npm-version is updating package.json with the new version number? Would it make sense to throw release in its own yaml restricted to push on master?

You've put your finger on a potential problem in the code here, I think. Currently it is not automatically updating the `version` attribute in package.json to match the version number that's been shipped to the registry. There must be a way to make this work but I haven't put the time to sort it out. Right now that would have to be updated manually prior to sending the release out with the same version number. We could make that extra feature a requirement before we merge, if you like.

hobbes7878 commented:
> Yes, that's basically it. Though it should stamp the version with whatever you pass into the release name in GitHub's web interface.

Ah, that's right. I forgot about the web release interface. Was thinking of this triggering on, e.g., `git tag 1.0.0` (which _I think_ it will??).

> The release event that qualifies to run the new, final step of the action should only get triggered for the default branch, I think. We want the tests that precede it to run for every push to every branch, so I wouldn't want to add a workflow-level branch restriction. **{{ Yep, agree. }}** I could imagine adding another statement to the if clause if you can imagine a circumstance where a tagged release would run for another branch.

So I see the `startsWith(github.ref, 'refs/tags')` in the new release job stops it from running on anything other than a tag. But if somebody created a PR with a tag on a branch, would that get through here?

> I tend not to use that approach because I like having the linter and test jobs act as hurdles to prevent broken code from shipping, and I prefer not to replicate those jobs in two different actions. 

👍

> There must be a way to make this work but I haven't put the time in to sort it out. Right now that would have to be updated manually prior to sending the release out with the same version number. We could make that extra feature a requirement before we merge, if you like.

Hmm.. I think this might be easy enough to just commit back to the repo. `npm version` is gonna bump package.json automatically, so we just push that 1-line change back, and because action pushes don't trigger subsequent actions, I _think_ we're in the clear??

Happy to help write this up. My only blocker is the first Q there about manual tags added on branches.

palewire commented:
1. I don't know what will happen if you do a manual tag with `git`. My habit has been to limit my usage to the Releases interface here on the web. I can test it on my demo repo and see what happens. 
2. Committing back to the repo should be doable, in the same manner as a scraper. Though I think that the tagged action checks out a detached head and it might require some kind of slick git repository maneuver to commit back to the default branch.  

I can put these two items on my homework list and get back to you.

palewire commented:
Your suspicion was correct. Manually creating a tag and pushing it a branch did slip past the `if` clause and trigger the release job. That's bad. I've pushed a patch. Here's are the receipts on the bug, which should not be resolved.

I'm still working on the second bit of homework.

## Before

```bash
# palewire @ archbook in ~/Code/npm-ci-test on git:main o [4:27:58] 
$ git status
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean

# palewire @ archbook in ~/Code/npm-ci-test on git:main o [4:27:59] 
$ git checkout -b test-tag
Switched to a new branch 'test-tag'

# palewire @ archbook in ~/Code/npm-ci-test on git:test-tag o [4:28:08] 
$ git tag 1.0.0

# palewire @ archbook in ~/Code/npm-ci-test on git:test-tag o [4:28:19] 
$ git push origin test-tag --tags
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
To https://github.com/palewire/npm-ci-test.git
 * [new branch]      test-tag -> test-tag
 * [new tag]         1.0.0 -> 1.0.0
 ```

![Fri Feb 10 04:29:40 AM EST 2023](https://user-images.githubusercontent.com/9993/218055720-065ea79f-7a42-42cb-a737-96b9f18e1254.jpg)
![Fri Feb 10 04:30:12 AM EST 2023](https://user-images.githubusercontent.com/9993/218055774-c7097d2a-3ac6-4d88-bd23-27720a3745a4.jpg)

## After

```
# palewire @ archbook in ~/Code/npm-ci-test on git:vet-tags o [5:04:01] C:128
$ git checkout -b vet-the-branch
Switched to a new branch 'vet-the-branch'

# palewire @ archbook in ~/Code/npm-ci-test on git:vet-the-branch o [5:04:14] 
$ git tag 0.0.999

# palewire @ archbook in ~/Code/npm-ci-test on git:vet-the-branch o [5:04:22] 
$ git push origin vet-the-branch --tags
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
To https://github.com/palewire/npm-ci-test.git
 * [new branch]      vet-the-branch -> vet-the-branch
 * [new tag]         0.0.999 -> 0.0.999
 ```

![Fri Feb 10 05:05:19 AM EST 2023](https://user-images.githubusercontent.com/9993/218063545-2e3887c0-9f14-430e-b2fa-da78706e481c.jpg)

palewire commented:
I believe I've handled the commit issue. The key trick was adding the contents permission to the job. This pull request is again ready for your review.

palewire commented:
I've added documentation explaining the release process this enables.

hobbes7878 reviewed (commented):
I _think_ we can save the `branch` and `vet` steps. Here's an alternative `if` for the whole `release` job:

```yaml
if: |
      github.event_name == 'push' &&
      startsWith(github.ref, 'refs/tags') &&
      endsWith(github.event.base_ref, github.event.repository.default_branch) &&
      github.repository == 'reuters-graphics/bluprint'
```

Line-by-line:
1.  Only pushes
2. Only release tags
3. Only events triggered from the repo's default branch (e.g., two values in that condition thingy will be `refs/heads/master` & `master`, respectively)
4. Only run on our repo (in case someone forks us, we don't try publishing from their master/main)

Tried it out real dirty on a separate repo and blocked the tags on branches and let through those on main.

Watcha think?

palewire commented:
Hey, if it works it's better. My only nit pick would be that a doctoring or two might be nice in something that long.

hobbes7878 commented:
K. Tried out a trim of your OG that simplifies a few bits and I think will apply well across most of our other node libs, too.

Here's the idea: https://github.com/reuters-graphics/npm-ci-test/blob/main/.github/workflows/ci.yaml

Checks I've done:
- [x] Runs tests on all pushes/PRs 
- [x] Runs release only on tags
- [x] Does not run release on branch tags
- [x] Does not run release on tags on forks

Open question from me, I'm not sure when we'd run this on a [`workflow_dispatch`](https://github.com/reuters-graphics/npm-ci-test/blob/main/.github/workflows/ci.yaml#L6). I'm guessing for integrating with _other_ CI tools, but there's no real harm in't, I suppose.

palewire commented:
Okay. I've tried to reconcile my patch to use your new `if` statement. I'm agnostic on the `workflow_dispatch`, though I do like it to run an ad hoc test run every once in a while. Happy to lose it that's what you'd prefer.

Small bugs asides, my only major question left is whether the packaging commands I've include will work with this package. Are you doing any rollup work or other building locally before you upload a distribution with `npm publish`? If so, that will likely need to be added yet.

hobbes7878 commented:
I shoulda said, normally we run build in a `prepublishOnly` [lifecycle script](https://docs.npmjs.com/cli/v9/using-npm/scripts#life-cycle-scripts), so it's covered when we run `npm publish`. 

In _this_ repo, that's also tied into running [tests](https://github.com/reuters-graphics/bluprint/blob/master/package.json#L46-L48), so yeah, probably a bit of cleanup to dedup for this new pattern.

palewire commented:
Okay. Am I right to take it that would mean reduce `prePublishOnly` to `npm run build`? Are you include any additional flags to publish that instruct it to upload the `dist` folder created by build?

hobbes7878 commented:
> Am I right to take it that would mean reduce prePublishOnly to npm run build? 

yep!

> any additional flags to npm publish that instruct it to upload the dist folder created by build?

The [`main` prop](https://github.com/reuters-graphics/bluprint/blob/master/package.json#L15) in package.json handles that one.

palewire commented:
Thank you. I believe I've made that edit to the package.json's `prePublishOnly` command.

hobbes7878 commented on docs/developing.md:
```diff
@@ -15,3 +15,73 @@ Be sure to write a test for your action in `test/actions/`.
+
+![Tag dropdown](https://raw.githubusercontent.com/reuters-graphics/bluprint/master/.github/images/releasing-name-tag.png)
+
+### 4. Name the release
+
+Enter the same number into the "Release title" box.
+
+![Name box](https://raw.githubusercontent.com/reuters-graphics/bluprint/master/.github/images/releasing-name-release.png)
```
You can actually skip this bit. #5 auto populates that title from the release number.

palewire replied:
Awesome. This must be a recent UI improvement from GitHub.

hobbes7878 commented on .github/workflows/test-and-release.yaml:
```diff
@@ -0,0 +1,116 @@
+        name: Run tests
+        run: yarn test
+
+      - id: upload-artifact
+        name: Upload release candidate
+        uses: actions/upload-artifact@v3
+        with:
+          name: release-candidate
+          path: ./dist/
```
Tinny nit, but still a little unclear what this is useful for if we never use the artifact.

palewire replied:
Sure. It's just there as "what if you want to take a peek" thing. I can cut it.

hobbes7878 commented on .github/workflows/test-and-release.yaml:
```diff
@@ -0,0 +1,116 @@
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
+
+      - id: commit
+        name: Commit
+        uses: stefanzweifel/git-auto-commit-action@v4
+        with:
+          commit_message: Updated version to ${{ github.ref_name }}
+          branch: ${{ github.event.repository.default_branch }}
```
Think we could get away without this dependency like I did [here](https://github.com/reuters-graphics/npm-ci-test/blob/main/.github/workflows/ci.yaml#L77-L83) if we add [a ref](https://github.com/reuters-graphics/npm-ci-test/blob/main/.github/workflows/ci.yaml#L52) to the checkout options. But maybe there's some other goodness this dude is giving us that I'm missing??

palewire replied:
If you've verified the git footwork can pull it off, sure, I'm good with it. I'll put that in.

palewire replied:
I've made that change. My only outstanding question is whether we can count on a naked `git push` to send changes to the default branch with this configuration.

hobbes7878 replied:
It did in testing, and I _think_ that's thanks to that [ref](https://github.com/reuters-graphics/npm-ci-test/blob/main/.github/workflows/ci.yaml#L52) option on checkout. But I suppose we could be double sure with:

```yaml
git push origin ${{ github.event.repository.default_branch }}
```

palewire replied:
If it works, it works. Cool config trick.

hobbes7878 commented on .github/workflows/test-and-release.yaml:
```diff
@@ -0,0 +1,116 @@
+        name: Install dependencies
+        run: yarn install --frozen-lockfile
+
+      - id: build
+        name: Build library
+        run: yarn build
```
_Technically_ not needed, b/c we call [build in test](https://github.com/reuters-graphics/bluprint/blob/master/package.json#L48). Benefit of doing those internal build calls in package.json as I see it, btw, is it makes this workflow work as template regardless whether our lib has a build step (many don't these days...).

palewire replied:
Yes, you are correct. I cut the build step.

hobbes7878 reviewed (approved):
Couple last details, six of one half a dozen, but otherwise this looks good!
